coreference/bert_stuff.py

```diff
@@ -0,0 +1,175 @@
+    overflow_idx=[]
+  
+    for i,j in enumerate(loc_mentions):
+        if j[1]>=4096:
+            overflow_idx.append([i, j[1] ])
+    
+    idx = [x[0] for x in overflow_idx]
+    # create a list for appending the overflowing indices
+    m_new  = []
+    for i, j in enumerate(overflow_idx):
+        bert_sent = bert_sentences[j[0] ]
+        m_new.append(m[j[0],  j[1]-4096+1:j[1]+1] )
+    #convert them into a tensor and pad to original max length of the entire batch 
+    m_new_tensor = torch.stack(m_new) 
+    m_new_tensor = F.pad(m_new_tensor,pad =(0, m.shape[1]-m_new_tensor.shape[1]), value=1) 
+
+    m[idx] = m_new_tensor
+
+    k = m == parallel_model.module.vals[0]
+    p = m == parallel_model.module.vals[1]
+
+    v = (k.int() + p.int()).bool()
+    nz_indexes = v.nonzero()[:, 1].reshape(m.shape[0], 2)
+
+    q = torch.arange(m.shape[1])
+    q = q.repeat(m.shape[0], 1)
+    
+    #create the  binary maps 
+    msk_0 = (nz_indexes[:, 0].repeat(m.shape[1], 1).transpose(0, 1)) <= q
+    msk_1 = (nz_indexes[:, 1].repeat(m.shape[1], 1).transpose(0, 1)) >= q
+
+    msk_0_ar = (nz_indexes[:, 0].repeat(m.shape[1], 1).transpose(0, 1)) < q
+    msk_1_ar = (nz_indexes[:, 1].repeat(m.shape[1], 1).transpose(0, 1)) > q
+
+    attention_mask_g = msk_0.int() * msk_1.int()
+    # get the tokens within 4096 after chunking 
+    input_ids = input_ids[:, :4096]
+    
+    attention_mask[:, 0] = 2
+    attention_mask[attention_mask_g == 1] = 2
+    attention_mask = attention_mask[:, :4096]
+    arg1 = msk_0_ar.int() * msk_1_ar.int()
+    arg1 = arg1[:, :4096]
+    arg1 = arg1.to(device)
+
+    #get the model outputs in batches
+    with torch.no_grad():
+        for i in tqdm(range(0, len(bert_sentences), batch_size), desc="generating "):
+
+            men_vectors = parallel_model(input_ids[i:i+batch_size], attention_mask[i:i+batch_size], arg1[i:i+batch_size])
+            all_vectors.extend(men_vectors)
+    
+    return all_vectors
+
+
+def generate_cdlm_embeddings(mention_map, vec_map_path , key_name ='bert_doc', num_gpus=4, batch_size=150, cpu=False):
+    """
+    Generate cdlm embeddings of the mentions in the cdlm format
+
+    Parameters
+    ----------
+
+    mention_map: dict
+        Mention map containing the documents with their mention IDs
+
+    vec_map_path: str
+        Path to store vector maps 
+        
+    key_name : str
+        Whether shorter sentences or longer whole documents with special mention tokens
+
+    num_gpus: int
+        Number of gpus to use
+
+    batch_size:
+        Batch size
+
+    cpu: bool
+        If we want to use cpu instead of gpus
+
+    Returns
+    -------
+    tensors
+        list of tensors, ie, the representations of the mentions
+    """
+    
+   
+    men_ids, bert_sentences = zip(*[(men_id, men[key_name]) for men_id, men in mention_map.items()])
+    config_file_path = '../cdlm/config_pairwise_long_reg_span.json'
+    config = pyhocon.ConfigFactory.parse_file(config_file_path)
```

Comment:
why again? didn't you intialize this at the beginning of file? remove one of them

Reply:
removed the one at the beginning(along with the import statements) and kept the one inside the function.